rewrite-java/src/main/java/org/openrewrite/java/InstanceOfPatternMatch.java

```diff
@@ -0,0 +1,334 @@
+                Set<J> validContexts = contexts.get(instanceOf);
+                for (Iterator<?> it = cursor.getPath(); it.hasNext(); ) {
+                    Object next = it.next();
+                    if (validContexts.contains(next)) {
+                        replacements.put(typeCast, instanceOf);
+                        break;
+                    } else if (root == next) {
+                        break;
+                    }
+                }
+            }
+        }
+
+        public boolean isEmpty() {
+            return replacements.isEmpty();
+        }
+
+        public J.InstanceOf processInstanceOf(J.InstanceOf instanceOf, Cursor cursor) {
+            @Nullable JavaType type = toJavaType((TypeTree) instanceOf.getClazz());
+            String name = patternVariableName(instanceOf, cursor);
+            J.InstanceOf result = instanceOf.withPattern(new J.Identifier(
+                    randomId(),
+                    Space.build(" ", emptyList()),
+                    Markers.EMPTY,
+                    name,
+                    type,
+                    null));
+
+            // update entry in replacements to share the pattern variable name
+            for (Map.Entry<J.TypeCast, J.InstanceOf> entry : replacements.entrySet()) {
+                if (entry.getValue() == instanceOf) {
+                    entry.setValue(result);
+                }
+            }
+            return result;
+        }
+
+        @Nullable
+        public J processTypeCast(J.TypeCast typeCast, Cursor cursor) {
+            J.InstanceOf instanceOf = replacements.get(typeCast);
+            if (instanceOf != null && instanceOf.getPattern() != null) {
+                String name = ((J.Identifier) instanceOf.getPattern()).getSimpleName();
+                TypedTree owner = cursor.firstEnclosing(J.MethodDeclaration.class);
+                owner = owner != null ? owner : cursor.firstEnclosingOrThrow(J.ClassDeclaration.class);
+                JavaType.Variable fieldType = new JavaType.Variable(null, Flag.Default.getBitMask(), name, owner.getType(), typeCast.getType(), emptyList());
+                return new J.Identifier(
+                        randomId(),
+                        typeCast.getPrefix(),
+                        Markers.EMPTY,
+                        name,
+                        typeCast.getType(),
+                        fieldType);
+            }
+            return null;
+        }
+    }
+
+    @Nullable
+    private static JavaType toJavaType(TypedTree typeTree) {
+        if (typeTree instanceof J.ArrayType) {
+            JavaType.Array result = new JavaType.Array(null, ((J.ArrayType) typeTree).getElementType().getType());
+            for (int i = 0; i < ((J.ArrayType) typeTree).getDimensions().size() - 1; i++) {
+                result = new JavaType.Array(null, result);
+            }
+            return result;
+        }
+        return typeTree.getType();
```

Comment:
Why is this whole method's implementation not just `typeTree.getType()`? I don't understand the processing you are doing on `typeTree`. I would expect that `J.ArrayType` for `String[][]` would return a corresponding `JavaType.Array` without requiring any post-processing. Is that not the case?

Reply:
Correct. The `getType()` method returns the array element type (e.g. `int`) for a `J.ArrayType`. that is what confused me all along and I intuitively suspected this to be wrong.

Reply:
This question will be addressed with #2713. In the scope of that issue this method can then hopefully be removed again.